aspects/counter.go

```diff
@@ -1,33 +1,91 @@
 package ginmon
 
-import "github.com/gin-gonic/gin"
+import (
+	"time"
+
+	"github.com/gin-gonic/gin"
+)
 
 // CounterHandler is a Gin middleware function that increments a
 // global counter on each request.
-func CounterHandler(counter *CounterAspect) gin.HandlerFunc {
-	return func(c *gin.Context) {
-		counter.Inc()
-		c.Next()
+func CounterHandler(ca *CounterAspect) gin.HandlerFunc {
+	return func(ctx *gin.Context) {
+		ca.Inc(ctx)
+		ctx.Next()
+		ca.IncCodes(ctx)
 	}
 }
 
 // CounterAspect stores a counter
 type CounterAspect struct {
-	Count int
+	internalRequestsSum  int
+	internalRequests     map[string]int
+	internalRequestCodes map[int]int
+	RequestsSum          int            `json:"request_sum_per_minute"`
+	Requests             map[string]int `json:"requests_per_minute"`
+	RequestCodes         map[int]int    `json:"request_codes_per_minute"`
+}
+
+// NewCounterAspect returns a new initialized CounterAspect object.
+func NewCounterAspect() *CounterAspect {
+	ca := &CounterAspect{}
+	ca.internalRequestsSum = 0
+	ca.internalRequests = make(map[string]int, 0)
+	ca.internalRequestCodes = make(map[int]int, 0)
+	return ca
+}
+
+// StartTimer will call a forever loop in a goroutine to calculate
+// metrics for measurements every d ticks. The parameter of this
+// function should normally be 1 * time.Minute, if not it will expose
+// unintuive JSON keys (requests_per_minute and
+// request_sum_per_minute).
+func (ca *CounterAspect) StartTimer(d time.Duration) {
+	timer := time.Tick(d)
+	go func() {
+		for {
+			<-timer
+			ca.reset()
+		}
+	}()
 }
 
-func (a *CounterAspect) Inc() {
-	a.Count++
+// Inc will increment internal counters that are not exposed. Counters
+// will be exposed if you call reset().
+func (ca *CounterAspect) Inc(ctx *gin.Context) {
+	ca.internalRequestsSum++
+	ca.internalRequests[ctx.Request.URL.Path]++
```

Comment:
Maybe I misunderstand how this is used, but it looks like this would be called on every request? Thus it doesn't seem thread safe like this.